File: python/features/steps/steps.py
from behave import given, then, when

from fhir_openehr_bridge import TranslationService

_service = TranslationService()


@given("a valid FHIR Patient JSON")
def step_valid_patient(context):
    context.fhir_json = (
        '{"resourceType":"Patient","id":"bdd-1",'
        '"name":[{"family":"Smith","given":["John"]}],'
        '"gender":"male","birthDate":"1980-05-15"}'
    )


@given("a FHIR Observation JSON")
def step_observation(context):
    context.fhir_json = '{"resourceType":"Observation","status":"final"}'


@given("a valid openEHR demographics composition")
def step_valid_composition(context):
    context.composition = {
        "archetypeNodeId": "openEHR-EHR-COMPOSITION.demographics.v1",
        "ehrStatus": {"subjectId": "bdd-2"},
        "demographics": {"familyName": "Smith", "givenName": "John", "gender": "male"},
    }


@given("an openEHR composition without demographics")
def step_composition_without_demographics(context):
    context.composition = {
        "archetypeNodeId": "openEHR-EHR-COMPOSITION.demographics.v1",
        "ehrStatus": {"subjectId": "bdd-3"},
        "demographics": None,
    }


@when("I translate it to openEHR")
def step_translate_to_openehr(context):
    context.result = _service.fhir_to_openehr(context.fhir_json)


@when("I translate it to FHIR")
def step_translate_to_fhir(context):
    context.result = _service.openehr_to_fhir(context.composition)


@then("the translation succeeds")
def step_succeeds(context):
    assert context.result.success, context.result.issues


@then("the translation fails")
def step_fails(context):
    assert not context.result.success


@then('the openEHR demographics family name is "{expected}"')
def step_family_name(context, expected):
    assert context.result.value["demographics"]["familyName"] == expected


@then("the result is a FHIR Bundle containing a Patient")
def step_bundle_with_patient(context):
    bundle = context.result.value
    assert bundle["resourceType"] == "Bundle"
    assert bundle["entry"][0]["resource"]["resourceType"] == "Patient"

File: nodejs/test/bdd/steps/steps.ts
import assert from 'assert';
import { Given, When, Then, Before } from '@cucumber/cucumber';
import { TranslationService } from '../../../src/translationService';
import { OpenEhrComposition, TranslationResult } from '../../../src/models';

interface World {
  fhirJson?: string;
  composition?: OpenEhrComposition;
  fhirResult?: TranslationResult<OpenEhrComposition>;
  bundleResult?: TranslationResult<string>;
}

const service = new TranslationService();
let world: World;

Before(function () {
  world = {};
});

Given('a valid FHIR Patient JSON', function () {
  world.fhirJson = JSON.stringify({
    resourceType: 'Patient',
    id: 'bdd-1',
    name: [{ family: 'Smith', given: ['John'] }],
    gender: 'male',
    birthDate: '1980-05-15',
  });
});

Given('a FHIR Observation JSON', function () {
  world.fhirJson = JSON.stringify({ resourceType: 'Observation', status: 'final' });
});

Given('a valid openEHR demographics composition', function () {
  world.composition = {
    archetypeNodeId: 'openEHR-EHR-COMPOSITION.demographics.v1',
    name: 'Demographics',
    language: 'en',
    territory: 'GB',
    category: 'persistent',
    ehrStatus: { subjectId: 'bdd-2', subjectNamespace: 'DEMOGRAPHIC', isQueryable: true, isModifiable: true },
    demographics: { familyName: 'Smith', givenName: 'John', gender: 'male', identifiers: [] },
  };
});

Given('an openEHR composition without demographics', function () {
  world.composition = {
    archetypeNodeId: 'openEHR-EHR-COMPOSITION.demographics.v1',
    name: 'Demographics',
    language: 'en',
    territory: 'GB',
    category: 'persistent',
    ehrStatus: { subjectId: 'bdd-3', subjectNamespace: 'DEMOGRAPHIC', isQueryable: true, isModifiable: true },
    // @ts-expect-error intentionally invalid for the scenario
    demographics: null,
  };
});

When('I translate it to openEHR', function () {
  world.fhirResult = service.fhirToOpenEhr(world.fhirJson as string);
});

When('I translate it to FHIR', function () {
  world.bundleResult = service.openEhrToFhir(world.composition as OpenEhrComposition);
});

Then('the translation succeeds', function () {
  const ok = world.fhirResult?.success ?? world.bundleResult?.success;
  assert.strictEqual(ok, true);
});

Then('the translation fails', function () {
  const ok = world.fhirResult?.success ?? world.bundleResult?.success;
  assert.strictEqual(ok, false);
});

Then('the openEHR demographics family name is {string}', function (expected: string) {
  assert.strictEqual(world.fhirResult?.value?.demographics.familyName, expected);
});

Then('the result is a FHIR Bundle containing a Patient', function () {
  const bundle = JSON.parse(world.bundleResult?.value as string);
  assert.strictEqual(bundle.resourceType, 'Bundle');
  assert.strictEqual(bundle.entry[0].resource.resourceType, 'Patient');
});
